### PR TITLE
[WIP] Add support for various k-NN model variants

### PR DESCRIPTION
**Pull request by speug**
This pull request introduces the k-NN model type to JKML. It should be noted that this PR is still a work in progress, and should not be merged before issues are addressed.

# Usage
## Basic modelling
`JKML -knn -repr [representation, one of [fchl, mbdf, fchl-kernel]] [-nometric] -train [trainset] -test [testset]`
Use the optional -nometric flag to skip metric learning and just use the Euclidean metric.

## Hyperparameters
Train hyperparameters and save to `path_to_hypercache`:
`JKML -knn -repr [representation, one of [fchl, mbdf, fchl-kernel]] [-nometric] -train [trainset] -hyper -varsout [path to hyper_cache]`

Use previously trained hyperparameters:
`JKML -knn -repr [representation, one of [fchl, mbdf, fchl-kernel]] [-nometric] -train [trainset] -test [testset] -hyper-cache [path to hyper_cache]`

# Features
- vanilla k-NN model with a number of representations supported
- k-NN with metric learning via [MLKR](https://contrib.scikit-learn.org/metric-learn/_modules/metric_learn/mlkr.html#MLKR)
- k-NN with induced kernel metric from the FCHL'18 kernel (requires compiling Fortran separately!)
- automated CV hyperparameter optimisation via [skopt](https://scikit-optimize.github.io/stable/) and hyperparameter loading for all of the above

# Requirements
Install the metric_learn library via
`pip install metric-learn`
and the scikit-optimize library via
`pip install scikit-optimize`

# Fortran compilation on Grendel
```
cd JKCS 2.1/JKML/src/fortran
ml load python/3.11.1
JKpython
f2py -m ffchl_vp_tree -c ffchl_kernel_types.f90 ffchl_kernels.f90 ffchl_module.f90 ffchl_kernel_distance.f90 fchl_vp_tree.f90 --fcompiler=gfortran --f90flags='-fopenmp' -lgomp
```

# Outstanding issues
- [ ] Python package management no updated to reflect new requirements
- [ ] Fortran compilation not automated for any of the supported environments
- [ ] Documentation lacking

### CODE REVIEW

**Comment by speug** [comment]
I should also note that I haven't really changed all of these files; it is merely that I have an autoformatter which formats code based on the [Black](https://black.readthedocs.io/en/stable/index.html) standard. The real contributions are in JKML/JKML.py, JKML/src/fortran/ and JKML/src/QKNN.py
